File: trading-market-making/typescript/step4-send-jet.ts
import {
  Connection,
  ComputeBudgetProgram,
  Transaction,
} from "@solana/web3.js";

const ENDPOINT = "https://your-endpoint.rpcpool.com";
const TOKEN = "your-token";

async function sendWithJet(
  serializedTransaction: Buffer,
  endpoint: string,
  token: string
): Promise<string> {
  // Use your Triton endpoint — Jet routing is handled automatically
  const connection = new Connection(`${endpoint}/${token}`, "confirmed");

  const signature = await connection.sendRawTransaction(
    serializedTransaction,
    {
      skipPreflight: true, // simulate separately via simulateTransaction() if needed
      maxRetries: 0,       // disable server-side retries; implement your own retry logic instead
    }
  );

  console.log("Sent:", signature);
  return signature;
}

function addPriorityFee(tx: Transaction, microLamportsPerCU: number): void {
  tx.add(
    ComputeBudgetProgram.setComputeUnitPrice({
      microLamports: microLamportsPerCU,
    })
  );
}

// Confirm the transaction landed
async function confirmTransaction(
  signature: string,
  endpoint: string,
  token: string
): Promise<void> {
  const connection = new Connection(`${endpoint}/${token}`, "confirmed");
  const statuses = await connection.getSignatureStatuses([signature]);
  const status = statuses.value[0];

  if (
    status?.confirmationStatus === "confirmed" ||
    status?.confirmationStatus === "finalized"
  ) {
    console.log("Landed:", signature);
  } else if (status?.err) {
    console.error("Failed:", status.err);
    // Re-fetch a recent blockhash, re-sign, and resend
  }
}

export { sendWithJet, addPriorityFee, confirmTransaction };

File: trading-market-making/rust/src/step4_send_jet.rs
use solana_client::rpc_client::RpcClient;
use solana_client::rpc_config::RpcSendTransactionConfig;
use solana_sdk::transaction::Transaction;

const ENDPOINT: &str = "https://your-endpoint.rpcpool.com";
const TOKEN: &str = "your-token";

fn send_with_jet(
    transaction: &Transaction,
    endpoint: &str,
    token: &str,
) -> anyhow::Result<String> {
    let rpc_url = format!("{}/{}", endpoint, token);
    let client = RpcClient::new(rpc_url);

    let signature = client.send_transaction_with_config(
        transaction,
        RpcSendTransactionConfig {
            skip_preflight: true,
            max_retries: Some(0), // disable server-side retries; implement your own retry logic
            ..Default::default()
        },
    )?;

    println!("Sent: {}", signature);
    Ok(signature.to_string())
}

fn main() {
    println!("See send_with_jet() — wire up a Transaction and call it with your endpoint.");
    println!("ENDPOINT: {}", ENDPOINT);
    println!("TOKEN: {}", TOKEN);
}
